src/cargo/ops/common_for_install_and_uninstall.rs

```diff
@@ -603,8 +609,21 @@ where
             vers
         };
         let dep = Dependency::parse_no_deprecated(name, vers_spec, source.source_id())?;
-        let deps = source.query_vec(&dep)?;
-        match deps.iter().map(|p| p.package_id()).max() {
+        let deps = (|| {
+            if let Some(vers_spec) = vers_spec {
+                if semver::VersionReq::parse(vers_spec).unwrap().is_exact() {
+                    let deps = source.query_vec(&dep)?;
+                    if deps.len() == 1 {
+                        return Ok(deps);
+                    }
+                }
+            }
+            if let NeedsUpdate::TryWithoutFirst = needs_update {
```

Comment:
This enum variant is weirdly tightly coupled to the style of source being used, but the args here already seem pretty tightly coupled to their caller styles. I tried a few different ways of factoring this out to avoid that, and none of them seemed great...